esp8266/scripts/neopixel.py

```diff
@@ -2,30 +2,44 @@
 # MIT license; Copyright (c) 2016 Damien P. George
 
 from esp import neopixel_write
+from slicetools import indices
+
 
 class NeoPixel:
-    def __init__(self, pin, n):
+    ORDER = (1, 0, 2, 3)
+
+    def __init__(self, pin, n, bpp=3):
         self.pin = pin
         self.n = n
-        self.buf = bytearray(n * 3)
+        self.bpp = bpp
+        self.buf = bytearray(n * bpp)
         self.pin.init(pin.OUT)
 
+    def _reorder(self, val):
+        for i in self.ORDER[:self.bpp]:
+            yield val[max(i, self.bpp - 1)]
+
     def __setitem__(self, index, val):
-        r, g, b = val
-        self.buf[index * 3] = g
-        self.buf[index * 3 + 1] = r
-        self.buf[index * 3 + 2] = b
+        if hasattr(index, 'start'):
+            def unroll(val=val):
+                for color in val:
+                    for byte in self._reorder(color):
+                        yield byte
+            self.buf[index] = bytes(unroll())
+            self.n = len(self.buf) // self.bpp
+            return
+        offset = index * self.bpp
+        self.buf[offset:offset + self.bpp] = bytes(self._reorder(val))
```

Comment:
This seems like a lot of overhead just to set a single pixel.  Here you need to allocate a generator function (which itself makes a copy of self.ORDER), then make a temporary bytes object, and also make a slice object for the store operation.  How about the following implementation:
```
if hasattr(index, 'start'):
    index = index.start or 0
else:
    val = (val,)
index *= self.bpp
for col in val:
    for pix in range(self.bpp):
        self.buf[index] = col[self.ORDER[pix]]
        index += 1
```